test/conversion_test.go

```diff
@@ -0,0 +1,794 @@
+	"github.com/google/go-cmp/cmp/cmpopts"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	knativetest "knative.dev/pkg/test"
+	"knative.dev/pkg/test/helpers"
+)
+
+var (
+	supportedV1ConversionFeatureGates = map[string]string{
+		"enable-api-fields":         "alpha",
+		"enable-tekton-oci-bundles": "true",
+	}
+
+	filterLabels        = cmpopts.IgnoreFields(metav1.ObjectMeta{}, "Labels")
+	filterAnnotations   = cmpopts.IgnoreFields(metav1.ObjectMeta{}, "Annotations")
+	filterTaskRunStatus = cmpopts.IgnoreFields(v1beta1.TaskRunStatusFields{}, "StartTime", "CompletionTime")
+
+	v1beta1TaskYaml = `
+metadata:
+  name: %s
+  namespace: %s
+spec:
+  steps:
+  - name: step
+    image: gcr.io/google.com/cloudsdktool/cloud-sdk:alpine
+    command: ['/bin/bash']
+    args: ['-c', 'gcloud auth activate-service-account --key-file /var/secret/bucket-secret/bucket-secret-key']
+    workingDir: /dir
+    env:
+    - name: MY_VAR1
+      value: foo
+    resources:
+      inputs:
+      - name: workspace
+        resource: source-repo
+      outputs:
+      - name: workspace
+        resource: source-repo
+    volumeMounts:
+      - name: messages
+        mountPath: /messages
+    imagePullPolicy: IfNotPresent
+    securityContext: 
+      runAsUser: 2000
+    timeout: 60s
+    secret:
+      secretName: test-ssh-credentials
+    onError: continue
+  stepTemplate:
+    image: gcr.io/google.com/cloudsdktool/cloud-sdk:alpine
+    command: ['/bin/bash']
+    env:
+    - name: QUX
+      value: original
+    args: ['-c', 'gcloud auth activate-service-account --key-file /var/secret/bucket-secret/bucket-secret-key']
+    workingDir: /dir
+    env:
+    - name: MY_VAR1
+      value: foo
+    resources:
+      inputs:
+      - name: workspaces
+        resource: source-repo
+      outputs:
+      - name: workspace
+        resource: source-repo
+    volumeMounts:
+    - name: messages
+      mountPath: /messages
+    imagePullPolicy: IfNotPresent
+    securityContext: 
+        runAsUser: 2000
+  sidecars:
+  - name: server
+    image: alpine/git:v2.26.2
+    command: ['/bin/bash']
+    args: ['-c', 'gcloud auth activate-service-account --key-file /var/secret/bucket-secret/bucket-secret-key']
+    workingDir: /dir
+    env:
+    - name: MY_VAR1
+      value: foo
+    resources:
+      inputs:
+      - name: workspace
+        resource: source-repo
+      outputs:
+      - name: workspace
+        resource: source-repo
+    volumeMounts:
+    - name: messages
+      mountPath: /messages
+    readinessProbe:
+      periodSeconds: 1
+    securityContext:
+      runAsUser: 0
+    volumeMounts:
+    - name: messages
+      mountPath: /messages
+    script: echo test
+  volumes:
+  - name: messages
+    emptyDir: {}
+  params:
+  - name: PARAM
+    description: param des
+    type: string
+    default: "1"
+  workspaces:
+  - name: workspace
+    description: description
+    mountPath: /foo
+    readOnly: true
+    optional: true
+  resources:
```

Comment:
I am confused seeing `resources` here 🙃

Reply:
`resources` is deprecated in v1beta1, so this needs to be covered by the conversion integration test to make sure it does not get dropped on the floor. 

For this specific case, on line #L612 it is being created by v1beta1TaskClient and converted to v1(which serializing the resources into annotations, which would correspond to the following comment at line #L158).

Please let me know if this makes any sense:)

Reply:
Thank you @JeromeJu 

Yup, I understand `resources` are deprecated and also, `resources` were not promoted to `beta`.

We also want to make sure `resources` are surfaced when we create V1 as part of the proposed migration [strategy](https://github.com/tektoncd/community/blob/main/teps/0074-deprecate-pipelineresources.md#upgrade--migration-strategy).

I understand they are being stored in `annotations` but aren't `annotations` part of the same API policies.

The only concern I have here is, when the users will notice pipeline resources in annotations as part of V1 CRDs, it might create a false expectation that the resources will be part of the V1 and not getting dropped anytime soon. I have been a huge fan of `resources` and would love that 🤣 

The pipeline resources were announced as deprecated in  `v0.30.0` and safe to remove since 12/20//2021.

https://github.com/tektoncd/pipeline/blob/main/docs/deprecations.md

From the policy perspective, I think, it's safe to delete the resources from the code base before we release V1 CRDs. Or at least drop them on the floor while converting to V1 but like you pointed out, it could lead to an incomplete/invalid CRD.

Reply:
I do understand the concerns here. So that to suggest that we need to make sure this is not misleading in the context of conversion_test.go. And I do agree that we would still need them in the examples here to be aligned with
https://github.com/tektoncd/pipeline/issues/4546. Therefore, would that makes sense to add the comments here in/ above the yamls where `resources` are referenced?

Reply:
Also wondering if this is out of scope for the current v1-CRD-release PR. The way dealing with the deprecated fields are discussed and addressed at https://github.com/tektoncd/pipeline/issues/4546. While I think in this PR what's added is the coverage for integration test of conversion, which shall not be the false expectation as this serves as a part of the [integration test ](https://github.com/tektoncd/pipeline/issues/5627).

Reply:
@jerop trying to understand, why not dropping `resources` on the floor while converting `beta` pipelines to `v1` instead adding them into `annotations`?

https://github.com/tektoncd/pipeline/pull/5331

FYI, no intention to block this as long as we comply with the policies.

My understanding is, the `annotations` approach is more applicable for the fields which are present in the higher versions e.g. `beta` but not in the lower versions e.g. `alpha`. Like `finally` and `when` expressions are introduced in `beta` CRDs but did not exist in `alpha` CRDs. These fields were serialized in `alpha` annotations to get the valid objects while converting up.

Reply:
@pritidesai I think one of the main intention (long-term) is to not lost *any* input from the user. 

In addition, we will support `v1beta1` and `v1` for a little while. As we do, we *will* support resource on `v1beta1` types, even when we switch the storage version to `v1`. If a user creates a `v1beta1` resource with a `PipelineResource`, it should work, as long as we support and serve `v1beta1` (or at least as long as the deprecation notice / 9 months thingy is). This means we don't really have any choice of dropping it on the floor if we want to keep our API policy promise.

Reply:
Thanks Vincent for chiming in and thanks Priti for the clarification question.

I think for the question it might be divided to 2 parts: 
- "whether we can drop `resources` on the floor":
   since `resources` is still in v1beta1 fields (though it has passed the time as you pointed in the deprecation.md to be removed), we might still have to support it.
- why going the `annotation` way
   I think the purpose for serializing into the `annotation` as in v1 there is no `resources` field and we do not want to apply it so that it can be converted back into v1beta1 before v1 gets to the stored version.

Reply:
@pritidesai we could do that but what about the next field that we deprecate ? The problem will be similar. 
Serializing into an annotation doesn't necessarily we support it "always". But it means we "keep" what the user sent us. If tomorrow we remove the `resources` field, then we would stop serializing, and for things where it was serialized it wouldn't be taken into account. As of today, someone who provide a `resources` in a `v1beta1` object has something working, if we release `v1` tomorrow and still serve `v1beta1`, it should still work for that user using `v1beta1` — there is nothing really more to it. The day we remove it to then we stop supporting it from `v1beta1`. In any case, today, a user using a `v1` wouldn't be able to user `resources` so it's relatively clear we are not supporting it.

Reply:
@JeromeJu and I have been looking through previous discussions, and although we discussed this quite a bit, it's actually not clear that we ever came to a decision. These alternatives [were proposed in API WG](https://docs.google.com/document/d/17PodAxG8hV351fBhSu7Y_OIPhGTVgj6OJ2lPphYYRpU/edit?disco=AAAAkU3A2Kg); [TEP-0105 to remove v1alpha1](https://github.com/tektoncd/community/blob/main/teps/0105-remove-pipeline-v1alpha1-api.md#proposal) doesn't set a proposal for removing PipelineResources, and [TEP-0074](https://github.com/tektoncd/community/blob/main/teps/0074-deprecate-pipelineresources.md#proposal) states that we can remove pipelineresources after 9 months, but not necessarily that we will. 

I believe we chose this approach after contributors expressed concerns about making sure that a v1 release doesn't include backwards incompatible changes to v1beta1, and I opened https://github.com/tektoncd/pipeline/issues/4546 for discussion. I didn't realize that this issue had not actually been decided; I'm sorry about that.

I agree with Vincent that the fact that you can't create a v1 CRD with resources makes it clear that we won't continue to support them, and prefer to keep the current approach. Would it be helpful for me to create an issue for removing PipelineResources completely, and we can target that for the next release?

Reply:
Thanks Priti, Vincent and Lee, added https://github.com/tektoncd/pipeline/issues/5785#issue-1460683293 for further discussion. Also added to the WG for next week.